src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -1552,5 +1552,64 @@ public void ChainTipAdvancesAt150Blocks_AlternativeChainPresented_RelevantBlockD
+        /// fork point at LC + 5, chain is not accepted.
+        /// </summary>
+        [Fact]
+        public void ChainAdvancesAfterACheckpoint_NewChainIsPresentedWithForkPoint_ChainIsNotAccepted()
+        {
+            // Chain header tree setup. Initial chain has 5 headers and checkpoint at h5.
+            // Example: h1=h2=h3=h4=(h5).
+            const int initialChainSize = 5;
+            TestContext ctx = new TestContextBuilder().WithInitialChain(initialChainSize).UseCheckpoints().Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader chainTip = ctx.InitialChainTip;
+
+            const int checkpointHeight = 5;
+            var checkpoint = new CheckpointFixture(checkpointHeight, chainTip.Header);
+            ctx.SetupCheckpoints(checkpoint);
+
+            // Setup max reorg to 100.
+            const int maxReorg = 100;
+            ctx.ChainStateMock.Setup(x => x.MaxReorgLength).Returns(maxReorg);
+
+            // Extend the chain with (checkpoint + MaxReorg + 10) headers, i.e. 115 headers.
+            const int extensionSize = 10;
+            const int chainASize = checkpointHeight + maxReorg + extensionSize;
+            ChainedHeader chainATip = ctx.ExtendAChain(chainASize, chainTip);
+
+            // Chain A is presented by peer 1.
+            List<BlockHeader> listOfChainABlockHeaders = ctx.ChainedHeaderToList(chainATip, chainASize);
+            ChainedHeader consumed = cht.ConnectNewHeaders(1, listOfChainABlockHeaders).Consumed;
+            ChainedHeader[] consumedChainAHeaders = consumed.ToArray(chainASize);
+            ChainedHeader[] originalChainAHeaders = chainATip.ToArray(chainASize);
+
+            // Sync all blocks from chain A.
+            for (int i = 0; i < chainASize; i++)
+            {
+                ChainedHeader currentChainTip = consumedChainAHeaders[i];
+                Block block = originalChainAHeaders[i].Block;
+
+                cht.BlockDataDownloaded(currentChainTip, block);
+                cht.PartialValidationSucceeded(currentChainTip, out bool fullValidationRequired);
+                cht.ConsensusTipChanged(currentChainTip);
+            }
+
+            // Create new chain B with 20 headers and a fork point at height 10.
+            const int forkPointHeight = 10;
+            const int chainBSize = 20;
+            ChainedHeader forkTip = chainATip.GetAncestor(checkpointHeight + forkPointHeight);
+            ChainedHeader chainBTip = ctx.ExtendAChain(initialChainSize + chainBSize - forkPointHeight, forkTip);
+
+            // Chain B is presented by peer 2.
+            List<BlockHeader> listOfChainBHeaders = ctx.ChainedHeaderToList(chainBTip, chainBSize);
+            ConnectNewHeadersResult connectNewHeadersResult = cht.ConnectNewHeaders(2, listOfChainBHeaders);
+
+            // Nothing is marked for download.
+            connectNewHeadersResult.DownloadFrom.Should().BeNull();
```

Comment:
The testable outcome here should be exception when the headers are attempted to be connected.